### ISSUE TITLE
[SMARTSCAN] Show PDF receipt thumbnails

### ISSUE THREAD

**Issue by trjExpensify** [Expensify/App#31432]
Coming from [here](https://github.com/Expensify/App/pull/30747#issuecomment-1802321587). CC: @sobitneupane @pradeepmdk @Gonals

If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.3.98-5
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @trjExpensify 
**Slack conversation:** https://github.com/Expensify/App/pull/30747#issuecomment-1802321587

## Action Performed:
1. Request Money > Scan > Upload a PDF receipt > choose a participant
2. Observe there isn't a preview of the PDF receipt in the thumbnail on the confirmation page
3. Click confirm to create the request
4. Observe there isn't a preview of the PDF receipt in the thumbnail on the report preview component in the chat
5. Click the report preview component 
6. Observe there isn't a preview of the PDF receipt in the thumbnail on the request preview component in the expense/iouReport
7. Click the request preview component
8. Observe there isn't a preview of the PDF receipt in the thumbnail on the receipt preview component in the request view

## Expected Result:
PDF receipts should show a preview in the relevant thumbnails throughout the app. 

P.s this also includes in the split preview component, so let's make sure it works there. 

## Actual Result:
Generic "empty" thumbnail previews are displayed for PDF receipts. 

## Workaround:
Yes, they can tap the thumbnail to see the PDF but they'd really not know to do that.

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Confirmation screen: 
<img width="377" alt="image" src="https://github.com/Expensify/App/assets/16232057/cb4400e9-3644-4add-89ff-9d823a5e4721">

Report preview on the chat: 
<img width="823" alt="image" src="https://github.com/Expensify/App/assets/16232057/1efe99bd-72ff-4b9d-9abf-aead949601c5">

Request preview on the expense/iouReport: 
<img width="826" alt="image" src="https://github.com/Expensify/App/assets/16232057/57cde704-3975-40df-a1cf-6f2574c1871a">

Receipt preview on the request: 
<img width="823" alt="image" src="https://github.com/Expensify/App/assets/16232057/6494ac25-5bdd-452f-9cbf-d8c0ee6931a4">

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by trjExpensify** [comment]
@0xmiroslav can you review this please? Also, @tienifr confirming in the results we're solving for all the thumbnails noted in the OP?

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to show PDF thumbnail in the process of uploading pdf receipt.

### What is the root cause of that problem?
The root cause of this issue is that, currently, we don't have an implementation of previewing PDF source by thumbnail. The PDF source here means expensify source like `https://www.expensify.com/chat-attachments/7486855856104260927/w_1c2aef0dea2f6e30acd3ca31c50286af09a4d3c9.pdf` or local file source starts with `blob://` or `file://`. 

Currently, uploading PDF attachment from chat can be previewed before upload and the PDF thumbnail is only available after it's been successfully uploaded and the backend will create an image source url for the PDF, like, `https://www.expensify.com/chat-attachments/7486855856104260927/w_1c2aef0dea2f6e30acd3ca31c50286af09a4d3c9.jpg.1024.jpg`

Intuitively, we should be able to preview the PDF thumbnail before uploading it to backend because we're able to preview PDF pages before uploading. 

The new feature to have a `ThumbnailPDF` which works on both server and local uri like what `Image` component supports is great complement to the App, especially for the use case like creating an IOU request and the file is not yet been uploaded to backend and the offline use case.

### What changes do you think we should make in order to solve the problem?
We can add an implementation of `ThumbnailPDF` which previews pdf uri (both server and local uri) in thumbnail. 
On the web platform, we can refer the implementation of `PDFView` by only rendering the first page of PDF file as the thumbnail. We can convert the first page to image, like this [example](https://github.com/charanvinay/PDF2Pic/blob/f0486eb9869bf19f40d63c2edb9de9cc6a3b9b18/src/Components/file-converter.js#L58).

On native platform, we can make use of the `singlePage` property of https://github.com/wonday/react-native-pdf.

This solution will need more effort to implement.

### What alternative solutions did you explore? (Optional)

**Comment by trjExpensify** [comment]
@0xmiroslav bump!

**Comment by trjExpensify** [comment]
Same. @0xmiroslav can you review these proposals today, please? They've been hanging out for 2 weeks.

**Comment by trjExpensify** [comment]
Perfect, thanks.

**Comment by trjExpensify** [comment]
@0xmiroslav did you review? 2 weeks is a pretty long time, and PDFs are pretty common receipt types. Thanks!

**Comment by trjExpensify** [comment]
>But do we need to show thumbnail for local pdf (before uploading) as well?

Yeah, we should show you a preview of the PDF like we do a preview of the local image attachment prior to upload.

**Comment by luacmartins** [comment]
@0xmiroslav should I review any specific proposal?

**Comment by luacmartins** [comment]
We simply replace the extension `.pdf` with `.1024.jpg`, so the resulting url is `filename.1024.jpg`. Is that what you needed?

**Comment by luacmartins** [comment]
I think we should show the first page as thumbnail, just like we do for regular pdf attachments. As for the ratio, I'm not sure what you mean. We should show the preview in the same format/dimensions we show other receipt file types, no?

**Comment by luacmartins** [comment]
Yes, we scale the thumbnail to fit a bounding box of 1024x1024 keeping its aspect ratio and only scaling down if the image is larger than 1024px.

**Comment by eh2077** [comment]
@0xmiroslav Can I have your feedback on my proposal https://github.com/Expensify/App/issues/31432#issuecomment-1820184229 ?

**Comment by luacmartins** [comment]
I'm not sure what you mean, if you already have `filename.1024.jpg` why would you need to replace the extension?

**Comment by trjExpensify** [comment]
Bump @eh2077 on @luacmartins's question!

**Comment by eh2077** [comment]
### Proposal
[Updated](https://github.com/Expensify/App/issues/31432#issuecomment-1820184229) proposal to use the new [thumbnail](https://github.com/wojtekmaj/react-pdf#thumbnail) feature of `react-pdf` on web.

@0xmiroslav I'll share a demo about the POC later.

**Comment by eh2077** [comment]
> I'm not sure what you mean, if you already have `filename.1024.jpg` why would you need to replace the extension?

@luacmartins When choosing a pdf receipt to scan from local disk, we're in the middle of flow and the pdf receipt is not yet uploaded to the backend, so in the `MoneyRequestConfirmPage`, we can't preview the pdf receipt in thumbnail, see the confirm screen shared by @trjExpensify 

<img width="200" src="https://github.com/Expensify/App/assets/117511920/61cad2ff-769b-4864-9956-9b542931d313">

**Comment by eh2077** [comment]
@trjExpensify Is it acceptable to show pdf receipt only when online? If so, then we can consider uploading the pdf to backend to get thumbnail before confirming the request, which will be easier to implemented though it's not perfect when offline.

**Comment by luacmartins** [comment]
@eh2077 we need the thumbnail to work offline too. 

>I guess the backend should have some logic to convert pdf to image, like this [SO](https://stackoverflow.com/questions/467793/how-do-i-convert-a-pdf-document-to-a-preview-image-in-php)

Yes, the backend uses ImageMagick to convert the pdf to jpg.

**Comment by trjExpensify** [comment]
>@eh2077 we need the thumbnail to work offline too.

Agreed 👍

**Comment by dylanexpensify** [comment]
@eh2077 any update here?

**Comment by dylanexpensify** [comment]
@0xmiroslav what's next here? Mind giving us an ETA?

**Comment by dylanexpensify** [comment]
@eh2077 @tienifr any chance you'd like to update proposals?

**Comment by trjExpensify** [comment]
Yeah, @tienifr @eh2077 we're super close on this one. Can you confirm in the next day or two before we look elsewhere?

**Comment by luacmartins** [comment]
Still waiting on proposals

**Comment by trjExpensify** [comment]
@eh2077 @tienifr, can you confirm if you'll be able to update in the next couple of days? Else @luacmartins, I think you might need to grab this one if that's okay with you. It's a HIGH issue and it's coming up in conversation with the real-world usage we have right now with the generalists policy and TU policy.

**Comment by trjExpensify** [comment]
👋 @0xmiroslav can you prioritise the review on this one, please? Thanks!

**Comment by trjExpensify** [comment]
Thanks so much! Somewhat related, but it seems like [loading the PDF](https://github.com/Expensify/App/issues/27680) in the receipt modal has broken again somehow, so the team approving PDF receipts can't see them at all now in /app so that's increased the urgency on this.

Asked a [question on the OG issue](https://github.com/Expensify/App/issues/27680#issuecomment-1865211719)for that about looking into it, but might be something easier to take care of here.

**Comment by AmjedNazzal** [comment]
@trjExpensify I haven't dug deep into the other issue you've mentioned, but from a quick glance I can see a couple of issues:
1) The url we pass to PDFView will look like this "/receipts/w_4904c6352b77bcf981b69797e5cde1c3dbc5e851.pdf" which will get automatically appended whatever we are running the app from, for example if we are running the dev build, the url will have dev appended in it and will look like this "https://dev.new.expensify.com:8082/receipts/w_4904c6352b77bcf981b69797e5cde1c3dbc5e851.pdf", meanwhile the backend specifies the url in transactions object to be "https://www.expensify.com/receipts/w_4904c6352b77bcf981b69797e5cde1c3dbc5e851.pdf" since we want to access the url in the actual expensify DB.

2) Even if I try correcting the url to "expensify.com", whenever pdfjs tries to access that, it will also fail which could mean there's also a problem in the BE side of things.

Would love to get insight from an engineer on why the actual url in onyx `transaction` is failing.

**Comment by luacmartins** [comment]
@AmjedNazzal do you see any console errors or request errors when trying to access the URL from pdfjs? Can you access the URL directly in your browser? [This url](https://www.expensify.com/receipts/w_4904c6352b77bcf981b69797e5cde1c3dbc5e851.pdf) works for me in the browser.

**Comment by AmjedNazzal** [comment]
@luacmartins Yes I am getting console error, even if I try this on prod which means I would be trying to access the url while logged in the correct user in prod and so the path will be correct, I get this console error: 
<img width="1280" alt="console error prod" src="https://github.com/Expensify/App/assets/74202751/8a4e67e2-ee2d-4218-bf25-e4f4f0c54f02">
And if I copy the url, it's actually the exact same url as the one I'm seeing in `transaction`, but if I try to open it in browser I get this: 
<img width="1280" alt="404 error" src="https://github.com/Expensify/App/assets/74202751/8f7ad64e-e95b-4012-a1ea-f9129e085f3b">

**Comment by trjExpensify** [comment]
I think Sobit figured out the regression here: https://github.com/Expensify/App/issues/27680#issuecomment-1866052293

**Comment by AmjedNazzal** [comment]
@0xmiroslav any updates? thank you!

**Comment by trjExpensify** [comment]
👋 @0xmiroslav this is a HIGH priority issue in wave5, can you prioritise this [proposal](https://github.com/Expensify/App/issues/31432#issuecomment-1863538914) review today please?

**Comment by eh2077** [comment]
> @eh2077 thumbnail in react-pdf v7 doesn't work when offline?

Sorry for delay just back from holiday. The new thumbnail feature should work but the v7 will break the pdf preview feature in the way we do in v6. I'll dig deeper and update soon.

**Comment by AmjedNazzal** [comment]
@0xmiroslav You are right that my original solution cannot handle native, to simplify and cleanup this and make code that can be overall useful in the future as well, we could create a component for thumbnails with the following steps: 

1) We would completely remove the logic for generating thumbnails and move it to a seperate new component called FileThumbnail and only include that within `MoneyTemporaryForRefactorRequestConfirmationList`: 
```javascript
{(receiptPath && receiptFilename && transaction) && (
<FileThumbnail
    transaction={transaction}
    receiptPath={receiptPath}
    receiptFilename={receiptFilename} 
/>
)} 
```

2) We then create the new component with index.js and index.native.js, in the index.js we use the original solution in my proposal: 
```javascript
function FileThumbnail ({transaction, receiptPath, receiptFilename}){
    ....
    let receiptImage;
    let receiptThumbnail;
    
    const [PDFReceiptImage, setPDFReceiptImage] = useState('');
    
    const generateThumbnail = () => {
        const offScreenCanvas = document.createElement('canvas');
        pdfjs.getDocument(transaction.receipt.source).promise
            .then(pdf => pdf.getPage(1))
            .then(page => {
                const viewport = page.getViewport({ scale: 1 });
                offScreenCanvas.width = viewport.width;
                offScreenCanvas.height = viewport.height;
                const context = offScreenCanvas.getContext('2d');
    
                page.render({ canvasContext: context, viewport }).promise
                .then(() => {
                    setPDFReceiptImage(offScreenCanvas.toDataURL('image/jpeg'));
                })
            })
            
    };
    
    if(receiptPath && receiptFilename && transaction){
        if ((Str.isPDF(transaction.filename || translate('attachmentView.unknownFilename')))) {
            generateThumbnail();
        }else {
            ({ image: receiptImage, thumbnail: receiptThumbnail } = ReceiptUtils.getThumbnailAndImageURIs(transaction, receiptPath, receiptFilename));
        }     
    }
    return (
        <>
        {(receiptImage || receiptThumbnail) && (
            <Image
                ....
            />
        )} 
        {PDFReceiptImage && (
            <Image 
                source={{ uri: PDFReceiptImage }}
                style={styles.moneyRequestImage} 
                isAuthTokenRequired={false} 
            />
        )}
        </>
    );
    
}
```

But in index.native.js, we can install and implement the use of `PdfThumbnail` from `react-native-pdf-thumbnail` which is a very simple library specifically made for this within the `generateThumbnail` function: 
```javascript
import PdfThumbnail from "react-native-pdf-thumbnail";
function FileThumbnail ({transaction, receiptPath, receiptFilename}){

....

 const generateThumbnail = async () => {
    const { uri } = await PdfThumbnail.generate(transaction.receipt.source, 0);
    setPDFReceiptImage(uri);
};

if(receiptPath && receiptFilename && receiptFile){
    if ((Str.isPDF(receiptFile.name || translate('attachmentView.unknownFilename')))) {
        generateThumbnail();
    }else {
        ({ image: receiptImage, thumbnail: receiptThumbnail } = ReceiptUtils.getThumbnailAndImageURIs(transaction, receiptPath, receiptFilename));
    }     
}

...

```
<img width="301" alt="pdfPreviewiOS" src="https://github.com/Expensify/App/assets/74202751/ddafae4c-a25d-4f88-ac2f-6da45d19e971">

**Comment by trjExpensify** [comment]
How is this linked to a deploy blocker? 🤔 

Separately, @0xmiroslav can you provide feedback on https://github.com/Expensify/App/issues/31432#issuecomment-1875909723

**Comment by trjExpensify** [comment]
>I am inclined to using thumbnail feature in react-pdf v7. So it makes sense to hold this for https://github.com/Expensify/App/pull/33083

Hm, I'm hesitant to add even further delay to fixing this as it's impacting a core flow, but if it's going to break again in a week I see the point. I think we need to up the urgency on getting https://github.com/Expensify/App/pull/33083 to the finish line though and treat it as a deliverable for this issue and #wave5 if we're now reliant on it. CC: @artus9033 & @jjcoffee to put this on your radar.

**Comment by trjExpensify** [comment]
Popped a hold in the title as that's getting finished.

**Comment by luacmartins** [comment]
Still on hold for https://github.com/Expensify/App/pull/33083

**Comment by trjExpensify** [comment]
Same, held on the linked PR.

**Comment by trjExpensify** [comment]
Still held, getting close though!

**Comment by eh2077** [comment]
@trjExpensify We don't need to consider the case of uploading a password protected pdf receipt right? Currently, uploading a protected pdf receipt results in uploading failure error.

**Comment by trjExpensify** [comment]
Sorry, not sure I'm quite following. Are you saying any upload of a password protected PDF is resulting in a failure? Maybe a quick screen recording would be helpful?

**Comment by eh2077** [comment]
Yes, here's the clip


https://github.com/Expensify/App/assets/117511920/547edfa3-1335-479e-bc4f-6e2444024b95

**Comment by trjExpensify** [comment]
Got it. If we don't allow password protected PDFs for receipt upload, we should be preventing that with the "file type not supported" alert modal. Remind me, do they have a special extension?

**Comment by eh2077** [comment]
No, the password protected PDFs have same `.pdf` extension as normal ones. I think we need to read them to know if they're password protected.

**Comment by eh2077** [comment]
> react-pdf is being updated to v7 in #33083 so you can research based on that codebase

### POC update
POC branch for testing: https://github.com/eh2077/App/commit/1ed6dbab2c14baa226312f5bc35d2f6901184b2a
1. Web implementation uses [Thumbnail](https://github.com/wojtekmaj/react-pdf?tab=readme-ov-file#thumbnail) of `react-pdf` v7
2. Native implementation uses the `singlePage` property of `react-native-pdf`


We can create a new component `PDFThumbnail` which is similar to the `ThumbnailImage` component. The `PDFThumbnail` can display a PDF source as a thumbnail. It works for both local file and remote source, online and offline modes.

POC videos 
1. Web online

https://github.com/Expensify/App/assets/117511920/909e1f9c-79fc-4eac-8420-6b88b0122826




2. Web offline


https://github.com/Expensify/App/assets/117511920/f7438222-6205-489f-9dd2-7d808ee8d8a9

4. Native 

| Android online  | Android offline |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/117511920/eb58b851-a120-465d-b433-949c471a07d3">  | <video src="https://github.com/Expensify/App/assets/117511920/4a78c0ea-0bd4-499e-95ec-144d8fea5faf">|

**Comment by eh2077** [comment]
@trjExpensify Can you help to check if the above demo videos look good to you overall?

**Comment by trjExpensify** [comment]
>No, the password protected PDFs have same .pdf extension as normal ones. I think we need to read them to know if they're password protected.

I see, okay. In which case, what's with the failed upload error? Granted, the receipt scanning will always fail to extract the receipt details because we can't see the content of the PDF to read it, but if the file extension is simply `.pdf`, we shouldn't be failing on upload. 

>@trjExpensify Can you help to check if the above demo videos look good to you overall?

Unfortunately, none of those vids are loading for me.

**Comment by mallenexpensify** [comment]
@s77rt reassigning, please take over as C+. If you don't have bandwidth, unassign yourself. Thanks

**Comment by trjExpensify** [comment]
FYI: PR for react-pdf is deployed to prod, so I took this off hold officially. 

Vids still don't load for me, but don't block on that if @s77rt & @luacmartins want to proceed and give this a review now.

**Comment by s77rt** [comment]
@tienifr Thanks for the proposal. Your RCA is correct. However your solution will only work if online and if the receipt is uploaded already.

**Comment by s77rt** [comment]
@eh2077 Thanks for the proposal. Your RCA is correct. The solution looks good to me. I have just one addition which is that we should always try to use the `.1024.jpg` thumbnail first since it's an actual image and faster to render.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/31432#issuecomment-1820184229)

**Comment by s77rt** [comment]
@AmjedNazzal Thanks for the proposal. Your RCA is correct. Your solution is close to what @eh2077 is suggesting. But I think it's better to use the the built in feature that `react-pdf` offers. Also I don't see the need for a new dependency `react-native-pdf-thumbnail`.

**Comment by luacmartins** [comment]
I agree with @s77rt's assessment and the videos looked good to me. Let's go ahead with it!

**Comment by dylanexpensify** [comment]
@eh2077 mind giving an ETA on when you'll have the PR raised?

**Comment by eh2077** [comment]
@dylanexpensify PR will be ready today

**Comment by eh2077** [comment]
Hi team,

As C+ @s77rt pointed out during PR reviewing, there're two issues need to clarify the expected behaviour

#### Issue 1: Uploading a password protected pdf results in a non-ending prompt unless you type the right pasword
See videos in https://github.com/Expensify/App/pull/35255#issuecomment-1913714726

See discussion https://github.com/Expensify/App/issues/31432#issuecomment-1904201073 and https://github.com/Expensify/App/issues/31432#issuecomment-1905771607 

#### Issue 2: The local preview do not match the one returned by the server
See videos in https://github.com/Expensify/App/pull/35255#issuecomment-1913713879

I feel it's hard to have the resolution of local PDF thumbnail same as the one returned by backend. See https://github.com/Expensify/App/issues/31432#issuecomment-1839330922 about the backend implementation.

cc @trjExpensify @luacmartins

**Comment by eh2077** [comment]
@trjExpensify Maybe you can try to refresh the page using Chrome if videos are not loading at the first time.

**Comment by trjExpensify** [comment]
@eh2077 what clarification questions do you have with regard to the expected behaviour?

**Comment by eh2077** [comment]
@trjExpensify For the case of uploading protected PDF, the expected behaviour is to show an alert modal right? As you mentioned earlier 

> Got it. If we don't allow password protected PDFs for receipt upload, we should be preventing that with the "file type not supported" alert modal. Remind me, do they have a special extension?

**Comment by trjExpensify** [comment]
>@trjExpensify For the case of uploading protected PDF, the expected behaviour is to show an alert modal right? As you mentioned earlier

Yeah, I was asking as I would have thought we could throw the file type error. Uploading a password protected PDF for receipt transcription is pointless, as we can't read any information from it, so I figured we might as well just prevent it.

**Comment by trjExpensify** [comment]
>For the second issue which is about the different resolution between local thumbnail and the one returned by backend. Is it acceptable to have the differences?

I'll defer to @luacmartins on this one. Ideally they are the same if we can do it.

**Comment by luacmartins** [comment]
@eh2077 why is it challenging to match the resolution?

**Comment by eh2077** [comment]
> @eh2077 why is it challenging to match the resolution?

@luacmartins Please find this https://github.com/Expensify/App/pull/35255#issuecomment-1929872013 FYI.

**Comment by youssef-lr** [comment]
Question, we already display previews when uploading a PDF attachment as a comment, why can't we use the same logic to display the PDF receipt's thumbnail?

**Comment by s77rt** [comment]
@youssef-lr We will be using that same logic however it only works while online since we need to wait for the server's response to get the pdf url (and thumbnail). If we create a scan request offline we need to mimic that behaviour which so far works well but it's not perfect e.g. [the preview image is not 100% the same](https://github.com/Expensify/App/pull/35255#issuecomment-1913713879).

**Comment by eh2077** [comment]
Hi team, here's the update of PR:

Improving error handling of uploading a protected PDF, see demo in  https://github.com/Expensify/App/pull/35255#issuecomment-1944087612

cc @trjExpensify

**Comment by dylanexpensify** [comment]
done!

**Comment by trjExpensify** [comment]
@CherylWalsh I'll ultimately defer to you, but in English for the "generic" file type not allowed we have this message: 

<img width="478" alt="image" src="https://github.com/Expensify/App/assets/16232057/e9c89e79-39a6-4d47-ab30-e2c7f6688bcb">

This is a special case in which a PDF file type is allowed, so it passes our front-end checks. At which point, we figure out it's a password protected PDF and throw the error. We can't SmartScan a PDF we can't access. So I think something like this would be fine and in-keeping with what we have: 

>**Attachment is the wrong type**
>Password protected PDF is not allowed

**Comment by trjExpensify** [comment]
CC: @iwiznia as I saw you raised this in the thread. Pending the above nod from Cheryl, I think the only portion we need a Spanish translation for then is "Password protected PDF", we already have the rest from the existing file type front-end validation error message.

**Comment by eh2077** [comment]
@CherylWalsh Friendly bump https://github.com/Expensify/App/issues/31432#issuecomment-1960648979

**Comment by CherylWalsh** [comment]
I feel the error message is too aggressive, I don't like "not allowed". It might be standard but I suggest this: 

**Invalid attachment type. Please avoid password-protected PDFs**

**Comment by dylanexpensify** [comment]
Ooo I like that copy for the error message, Cheryl!

**Comment by trjExpensify** [comment]
Can we suggest in the format of the alert modal? 

>**bold title**
>Second copy line

If we're going to make the password-protected PDF one read differently, we should update the front-end validation error for unsupported file types as well so they're consistent. Based on the suggestion that would be:

(PP-PDF)

>**Invalid attachment type**
>Please avoid password-protected PDFs

AND 

(Generic)

>**Invalid attachment type**
>Please avoid this file type


Yeah, not sure if those are better tbh. 🤔 

- "File type" is the correct term, I don't think "attachment type" is a term used anywhere commonly to refer to the extension of a file. 
- Some alternatives to "not allowed" commonly used might be "not permitted" or "not supported", perhaps? It's a pretty transactional error message at the end of the day. Maybe these below? 

(PP-PDF specific)

**Invalid file type**
>Password protected PDF is not supported

AND 

(Generic)

**Invalid file type**
>This file type is not supported

**Comment by CherylWalsh** [comment]
Yes, love this. If not attachment type then file type works. And I like 'not supported' instead of allowed. I know it's only a transactional error message but it does matter. 

(PP-PDF specific)

**Invalid file type**
Password-protected PDF is not supported          (it's hyphenated)

AND

(Generic)

**Invalid file type**
This file type is not supported

**Comment by trjExpensify** [comment]
Perfect. So next steps: 

1. @eh2077 can you make sure to include updating the generic file type error message in your PR, so we keep things consistent. 
2. @iwiznia can we get Spanish translations from you for the below, please? 

>**Invalid file type**
>Password-protected PDF is not supported

AND

>**Invalid file type**
>This file type is not supported

**Comment by iwiznia** [comment]
`This file type is not supported` we kind of already have this: https://github.com/Expensify/App/blob/b336cd3b4c258d23071fb47305bc279c7adac477/src/languages/en.ts#L337 and this https://github.com/Expensify/App/blob/b336cd3b4c258d23071fb47305bc279c7adac477/src/languages/en.ts#L445, should we reuse and unify those?

Invalid file type: Tipo de archivo inválido
Password-protected PDF is not supported: Los PDFs con password no son compatibles

**Comment by trjExpensify** [comment]
>This file type is not supported we kind of already have this:

Yeah, the idea is to replace `This file type is not allowed` with `This file type is not supported` as part of this issue now to align them.

**Comment by iwiznia** [comment]
ok you already have the text and translation for that

**Comment by trjExpensify** [comment]
So for the avoidance of doubt and @eh2077 benefit, the below are the final set. 

(PP-PDF specific)

>**Invalid file type**
>Password-protected PDF is not supported

>**Tipo de archivo inválido** 
>Los PDFs con password no son compatibles

(Generic)

>**Invalid file type**
>This file type is not allowed

>**Tipo de archivo inválido** 
>Este tipo de archivo no son compatibles

**Comment by iwiznia** [comment]
This is wrong:
> Este tipo de archivo no son compatibles

Should be:
> Este tipo de archivo no es compatible

**Comment by trjExpensify** [comment]
Lol, got it. Updated.

**Comment by s77rt** [comment]
PR has been deployed to staging 2 hours ago

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: Not a a bug. This is a feature request
- The offending PR has been commented on: n/a
- A discussion in #expensify-bugs has been started: n/a
- Determine if we should create a regression test for this bug: Yes

___

**Regression Test Proposal**
1. Go to any chat and request money
2. Choose Scan tab and upload a pdf file
3. Verify that you're able to preview the PDF receipt in the confirmation page
4. Complete the money request process
5. Verify that you're able to preview the PDF receipt in the chat report
6. Click on the PDF preview
7. Verify that you're able to preview the PDF receipt in the IOU report
8. Click on the PDF preview
9. Verify that you're able to preview the PDF receipt in the money request report
10. Click on the PDF preview
11. Verify that you are displayed the PDF file

**Comment by trjExpensify** [comment]
Hey @eh2077, thanks for the request. I agree the scope of this issue increased to encompass password-protected PDFs and updating the file type modal. I've added a bonus and sent you both offers for $750.

**Comment by s77rt** [comment]
@trjExpensify We had a regression https://github.com/Expensify/App/issues/37852. Should the payment here be $375 in that case?

**Comment by trjExpensify** [comment]
Ah, I didn't see that. Okay cool, so then I'm fine with keeping the bounty at $500 and not reducing for that regression. Once you guys accept the offers, I'll pay a different amount. 👍

**Comment by s77rt** [comment]
@trjExpensify Accepted! Thank you!

**Comment by trjExpensify** [comment]
Paid both of you, thanks!
